### PR TITLE
Change upload value to contain Python types

### PR DESCRIPTION
**Pull request by pbugnion**
Prior to this PR, the `.value` traitlet for the `FileUpload` widget was a list of dictionaries. Each dictionary contained the raw metadata generated by the browser, as well as the content as a memoryview.

This PR changes the return value to be a tuple of `Bunch` instances. Attribute names inside the bunch are converted to snake case and the last modified is converted to a zoned datetime.

For instance:

```python
# f is a FileUpload instance
>>> f.value
({'name': 'some-name.png',
  'type': 'image/png',
  'size': 29841,
  'content': <memory at 0x118001940>,
  'last_modified': 1578491179205},)
```

This PR also removes syncing an 'error' attribute per file. This was not exposed in the kernel, and it was never set to anything apart from an empty string in the frontend. Note that the global error attribute remains.

Note that this PR has some overlap with #2715 : both share datetime serialization / deserialization considerations. In here, I have done the least amount of work -- I just send a timestamp over the wire, and manage conversion in the kernel. Once 2548 is merged, it may be worth overhauling serialization to have a common interface. That won't be a breaking change.

Closes issue #2727 .

### CODE REVIEW

**Comment by pbugnion** [comment]
This is now ready for review.

File: ipywidgets/widgets/widget_upload.py [hunk 1]
```diff
@@ -16,12 +17,107 @@
 from .widget_button import ButtonStyle
 from .widget import register, widget_serialization
 from .trait_types import InstanceDict
+from traitlets.utils.bunch import Bunch
```
Comment (jasongrout) [inline ipywidgets/widgets/widget_upload.py]: I think you can just do `from traitlets import Bunch` to not rely on the internal traitlets structure.

File: ipywidgets/widgets/widget_upload.py [hunk 2]
```diff
@@ -16,12 +17,107 @@
...
+
+def _deserialize_value(js, _):
+    return [_deserialize_single_file(entry) for entry in js]
+
+
+def _serialize_single_file(uploaded_file):
+    js = {}
+    for attribute in ['name', 'type', 'size', 'content']:
+        js[attribute] = uploaded_file[attribute]
+    js['lastModified'] = int(uploaded_file['last_modified'].timestamp() * 1000)
```
Comment (jasongrout) [inline ipywidgets/widgets/widget_upload.py]: Our general convention, at least in default controls, has been snake_case for the attributes in the protocol messages. Should we keep it snake_case here, as `last_modified`?

File: ipywidgets/widgets/widget_upload.py [hunk 3]
```diff
@@ -16,12 +17,107 @@
...
+
+    style: widgets.widget_button.ButtonStyle, optional
+        Style configuration for the button.
+
+    value: List[Dict], optional
+        The value of the last uploaded file or set of files. See the
+        documentation for details of how to use this to retrieve file
+        content and metadata:
+        https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html#File-Upload
+    error: str, optional
```
Comment (jasongrout) [inline ipywidgets/widgets/widget_upload.py]: Missing blank line

File: ipywidgets/widgets/widget_upload.py [hunk 4]
```diff
@@ -32,10 +128,11 @@ class FileUpload(DescriptionWidget, ValueWidget, CoreWidget):
     icon = Unicode('upload', help="Font-awesome icon name, without the 'fa-' prefix.").tag(sync=True)
     button_style = CaselessStrEnum(
         values=['primary', 'success', 'info', 'warning', 'danger', ''], default_value='',
-        help="""Use a predefined styling for the button.""").tag(sync=True)
+        help='Use a predefined styling for the button.').tag(sync=True)
     style = InstanceDict(ButtonStyle).tag(sync=True, **widget_serialization)
     error = Unicode(help='Error message').tag(sync=True)
-    value = List(Dict(), help="The file upload value").tag(sync=True)
+    value = List(Dict(), help='The file upload value').tag(
```
Comment (jasongrout) [inline ipywidgets/widgets/widget_upload.py]: Let's make this a TypedTuple so it is read-only.

**Comment by jasongrout** [comment]
This looks great! A couple of inline comments above, and then I think it's ready to go.

File: ipywidgets/widgets/widget_upload.py (hunk 2 above)
Comment (jasongrout) [reply ipywidgets/widgets/widget_upload.py]: Another option is to just call it `modified`, or even `mtime`, which mirrors the `stat()` structure and how you address it in python and presumably lots of other languages that let you make a stat call.

Comment (pbugnion) [reply ipywidgets/widgets/widget_upload.py]: `mtime` sounds good.

**Comment by pbugnion** [comment]
Thanks for the feedback. I made all the changes discussed.

**Comment by jasongrout** [comment]
Thanks! This looks really great.

**Pull request merged**